// @azure/identity/dist-esm/src/credentials/interactiveBrowserCredential.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
import { __awaiter } from "tslib";
import { credentialLogger, formatError, formatSuccess } from "../util/logging";
import { DefaultTenantId, DeveloperSignOnClientId } from "../constants";
import { AuthenticationRequired, MsalClient } from "../client/msalClient";
import express from "express";
import open from "open";
import { checkTenantId } from "../util/checkTenantId";
const logger = credentialLogger("InteractiveBrowserCredential");
/**
 * Enables authentication to Azure Active Directory inside of the web browser
 * using the interactive login flow, either via browser redirects or a popup
 * window.  This credential is not currently supported in Node.js.
 */
export class InteractiveBrowserCredential {
    constructor(options) {
        const tenantId = (options && options.tenantId) || DefaultTenantId;
        const clientId = (options && options.clientId) || DeveloperSignOnClientId;
        checkTenantId(logger, tenantId);
        // const persistenceEnabled = options?.persistenceEnabled ? options?.persistenceEnabled : false;
        // const authenticationRecord = options?.authenticationRecord;
        if (options && options.redirectUri) {
            if (typeof options.redirectUri === "string") {
                this.redirectUri = options.redirectUri;
            }
            else {
                this.redirectUri = options.redirectUri();
            }
        }
        else {
            this.redirectUri = "http://localhost";
        }
        const url = new URL(this.redirectUri);
        this.port = parseInt(url.port);
        if (isNaN(this.port)) {
            this.port = 80;
        }
        let authorityHost;
        if (options && options.authorityHost) {
            if (options.authorityHost.endsWith("/")) {
                authorityHost = options.authorityHost + tenantId;
            }
            else {
                authorityHost = options.authorityHost + "/" + tenantId;
            }
        }
        else {
            authorityHost = "https://login.microsoftonline.com/" + tenantId;
        }
        this.msalClient = new MsalClient({ clientId, authority: authorityHost }, false, undefined, options);
    }
    /**
     * Authenticates with Azure Active Directory and returns an access token if
     * successful.  If authentication cannot be performed at this time, this method may
     * return null.  If an error occurs during authentication, an {@link AuthenticationError}
     * containing failure details will be thrown.
     *
     * @param scopes The list of scopes for which the token will have access.
     * @param options The options used to configure any requests this
     *                TokenCredential implementation might make.
     */
    getToken(scopes, _options) {
        const scopeArray = typeof scopes === "object" ? scopes : [scopes];
        return this.msalClient.acquireTokenFromCache(scopeArray).catch((e) => {
            if (e instanceof AuthenticationRequired) {
                return this.acquireTokenFromBrowser(scopeArray);
            }
            else {
                logger.getToken.info(formatError(scopes, e));
                throw e;
            }
        });
    }
    openAuthCodeUrl(scopeArray) {
        return __awaiter(this, void 0, void 0, function* () {
            const authCodeUrlParameters = {
                scopes: scopeArray,
                redirectUri: this.redirectUri
            };
            const response = yield this.msalClient.getAuthCodeUrl(authCodeUrlParameters);
            yield open(response);
        });
    }
    acquireTokenFromBrowser(scopeArray) {
        return __awaiter(this, void 0, void 0, function* () {
            // eslint-disable-next-line
            return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
                // eslint-disable-next-line
                let listen;
                let socketToDestroy;
                function cleanup() {
                    if (listen) {
                        listen.close();
                    }
                    if (socketToDestroy) {
                        socketToDestroy.destroy();
                    }
                }
                // Create Express App and Routes
                const app = express();
                app.get("/", (req, res) => __awaiter(this, void 0, void 0, function* () {
                    const tokenRequest = {
                        code: req.query.code,
                        redirectUri: this.redirectUri,
                        scopes: scopeArray
                    };
                    try {
                        const authResponse = yield this.msalClient.acquireTokenByCode(tokenRequest);
                        const successMessage = `Authentication Complete. You can close the browser and return to the application.`;
                        const expiresOnTimestamp = authResponse === null || authResponse === void 0 ? void 0 : authResponse.expiresOn.valueOf();
                        res.status(200).send(successMessage);
                        logger.getToken.info(formatSuccess(scopeArray));
                        resolve({
                            expiresOnTimestamp,
                            token: authResponse.accessToken
                        });
                    }
                    catch (error) {
                        const errorMessage = formatError(scopeArray, `${req.query["error"]}. ${req.query["error_description"]}`);
                        res.status(500).send(errorMessage);
                        logger.getToken.info(errorMessage);
                        reject(new Error(errorMessage));
                    }
                    finally {
                        cleanup();
                    }
                }));
                listen = app.listen(this.port, () => logger.info(`Msal Node Auth Code Sample app listening on port ${this.port}!`));
                listen.on("connection", (socket) => (socketToDestroy = socket));
                try {
                    yield this.openAuthCodeUrl(scopeArray);
                }
                catch (e) {
                    cleanup();
                    throw e;
                }
            }));
        });
    }
}
